### PR TITLE
[docs] Add sampling docs and improve distributed tracing

### PR DESCRIPTION
**Pull request by bmorelli25**
## Motivation/summary

New sampling documentation aims to address these questions:
* What is sampling?
* Why sample?
* How does sampling impact distributed tracing?
* What data is sampled?
* How can I adjust the sampling rate?

New distributed tracing documentation aims to address these questions:
* What is DT and why is it important?
* How does trace propagation work + examples
* How can I manually propagate trace context with each agent's API

## Docs preview

Previews! Get your previews here!

* **[New sampling page](https://apm-server_4475.docs-preview.app.elstc.co/guide/en/apm/get-started/master/trace-sampling.html)**
* **[Enhanced DT page](https://apm-server_4475.docs-preview.app.elstc.co/guide/en/apm/get-started/master/distributed-tracing.html)**

## Related issues

Closes https://github.com/elastic/apm/issues/357
Closes https://github.com/elastic/apm/issues/221
Closes https://github.com/elastic/apm/issues/248
Closes https://github.com/elastic/apm-server/issues/2062

### CODE REVIEW

**Comment by bmorelli25** [comment]
@elasticmachine, run elasticsearch-ci/docs

File: docs/tab-widgets/distributed-trace-receive.asciidoc [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+2. Start a new transaction or span as a child of the incoming transaction with
+{apm-go-ref}/api.html#tracer-api-start-transaction-options[`StartTransactionOptions`] or
+{apm-go-ref}/api.html#transaction-start-span-options[`StartSpanOptions`]
+
+Example starting a new transaction:
+
+[source,go]
+----
+// Receive incoming TraceContext
```
Comment (axw) [inline docs/tab-widgets/distributed-trace-receive.asciidoc:15]:
```suggestion
// Receive incoming TraceContext
traceContext, _ := apmhttp.ParseTraceparentHeader(r.Header.Get("Traceparent"))
traceContext.State, _ = apmhttp.ParseTracestateHeader(r.Header["Tracestate"]...)
```

File: docs/tab-widgets/distributed-trace-send.asciidoc [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+2. Get the active TraceContext.
+
+3. Send the TraceContext to the receiving service.
+
+[source,go]
+----
+transaction := apm.DefaultTracer.StartTransaction("GET /", "request")
+// Get TraceContext from current Transaction
+// Send TraceContext to receiving service
```
Comment (axw) [inline docs/tab-widgets/distributed-trace-send.asciidoc:17]:
```suggestion
// Get TraceContext from current Transaction
traceContext := transaction.TraceContext()

// Send TraceContext to receiving service
traceparent := apmhttp.FormatTraceparentHeader(traceContext))
tracestate := traceContext.State.String()
```

File: docs/tab-widgets/distributed-trace-receive.asciidoc (hunk 1 above)
Comment (axw) [reply docs/tab-widgets/distributed-trace-receive.asciidoc:15]: In this sample code, `r` refers to a [net/http.Request](https://golang.org/pkg/net/http/#Request) being handled by a web service.

File: docs/guide/trace-sampling.asciidoc [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+the APM app always knows which transactions have and haven't been sampled.
+This prevents the app from showing broken traces.
+In addition, because transaction and error data is never sampled,
+you can always expect metrics and errors to be accurately reflected in the APM app.
+
+*Service maps*
+
+Service maps rely on distributed traces to draw connections between services.
+A minimum required version of APM agents is required for Service maps to work.
+See {kibana-ref}/service-maps.html[Service maps] for more information.
```
Comment (bmorelli25) [inline docs/guide/trace-sampling.asciidoc:79]: @ogupte Can you confirm this is accurate/make any recommendations for improvement?

Please also take a look at [L33-38](https://github.com/elastic/apm-server/pull/4475/files#diff-e537cc42e5aaa8d95f8ef9b9f1e7e468d74c43e0d7869a2973789d83bf96c1c4R33-R38).

File: docs/tab-widgets/distributed-trace-receive.asciidoc [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+# how to decode `trace_context`?
+
+ElasticAPM.with_transaction 'Do things' do
+  ElasticAPM.start_span 'Do one of the things' # How do we add `trace_context` to this?
+   # ... do something
+  ElasticAPM.end_span
+end
+----
+
+// end::ruby[]
```
Comment (bmorelli25) [inline docs/tab-widgets/distributed-trace-receive.asciidoc:162]: @elastic/apm-agent-ruby -- I'm having trouble figuring things out for the Ruby agent. Can you please take a look at this example and make recommendations?

These docs are for manually propagating the traceparent in a sending service

File: docs/tab-widgets/distributed-trace-send.asciidoc [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+[source,ruby]
+----
+ElasticAPM.start_transaction('Name') <1>
+# Get the trace_context and save to variable
+# Send the trace_context to another service
+----
+<1> Start a transaction
+
+// end::ruby[]
```
Comment (bmorelli25) [inline docs/tab-widgets/distributed-trace-send.asciidoc:162]: @elastic/apm-agent-ruby (part two)

These docs are for manually propagating the traceparent in a **receiving** service

File: docs/guide/trace-sampling.asciidoc (hunk 3 above)
Comment (ogupte) [reply docs/guide/trace-sampling.asciidoc:79]: Looks good to me!

File: docs/tab-widgets/distributed-trace-receive.asciidoc (hunk 4 above)
Comment (mikker) [reply docs/tab-widgets/distributed-trace-receive.asciidoc:162]: > Instantiate your own trace context from a Rack `env` object, then pass it when creating your transaction:

```ruby
context = ElasticAPM::TraceContext.parse(env: env)
ElasticAPM.with_transaction("Do things", trace_context: context) do
  ElasticAPM.with_span("Do nested thing", trace_context: context) do
    # However, spans will auto-create a child to their parents transaction's
    # TraceContext of none is passed
  end
end
```

Looking on the other languages' docs, I think this part of the docs is for _receiving_ services, though?

File: docs/tab-widgets/distributed-trace-send.asciidoc (hunk 5 above)
Comment (mikker) [reply docs/tab-widgets/distributed-trace-send.asciidoc:162]: 1. Start a span
1. Get its TraceContext and use it to generate a Traceparent header

```ruby
ElasticAPM.with_span "Name" do |span|
  header = span.trace_context.traceparent.to_header
  # etc ...
end
```

(This is a _sending_ service, see above)

File: docs/tab-widgets/distributed-trace-receive.asciidoc (hunk 4 above)
Comment (bmorelli25) [reply docs/tab-widgets/distributed-trace-receive.asciidoc:162]: Oops! You're right. Thanks!

**Comment by bmorelli25** [comment]
This PR is (finally) ready for review. I went wild with the reviewer list, so don't feel obligated to review — unless you're on the @elastic/obs-docs team, then yes, you are obligated 😄.

File: docs/tab-widgets/distributed-trace-receive.asciidoc [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+Example:
+
+[source,go]
+----
+// Receive incoming TraceContext
+traceContext, _ := apmhttp.ParseTraceparentHeader(r.Header.Get("Traceparent")) <1>
+traceContext.State, _ = apmhttp.ParseTracestateHeader(r.Header["Tracestate"]...) <2>
+
+opts := apm.TransactionOptions{
+	Start: time.Now(),
```
Comment (axw) [inline docs/tab-widgets/distributed-trace-receive.asciidoc:22]:
```suggestion
```

The current time will be used if Start is unspecified. Seeing as this example is about context propagation, I'd leave this out to minimise distractions.

**Review by axw — APPROVED** [review]
Looks great! Just one minor suggestion for the Go examples.

**Review by EamonnTP — APPROVED** [review]
Looks great!

File: docs/tab-widgets/distributed-trace-receive.asciidoc [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+Start a new transaction or span as a child of the incoming transaction or span with
+{apm-ruby-ref}/api.html#api-agent-with_transaction[`with_transaction`] or
+{apm-ruby-ref}/api.html#api-agent-with_span[`with_span`].
+
+Example:
+
+[source,ruby]
+----
+context = ElasticAPM::TraceContext.parse(env: env) <1>
```
Comment (bmorelli25) [inline docs/tab-widgets/distributed-trace-receive.asciidoc:152]:
```suggestion
# env being a Rack env
context = ElasticAPM::TraceContext.parse(env: env) <1>
```

**Pull request merged**